### PR TITLE
Feature/10123 add subscription expire section

### PR DESCRIPTION
**Pull request by JorgeMucientes**
Closes: #10147

### Description

Adds a new section to product detail screen enabling the user to edit the `length` of the subscription product. Note that this PR doesn't include the changes needed to support this in a product variation. That support is added in this next PR: https://github.com/woocommerce/woocommerce-android/pull/10153 to reduce the scope of this one. 

### Testing instructions
Shamelessly based/copied from this PR #10138 😅 
##### Subscription
1. Make sure that the Woo Subscriptions plugin is installed in your store.
4. Create a Subscription product.
5. Open the app, then navigate to this product.
6. Click on "Subscription expiration" row
7. Change the subscription length.
8. Go back to the product, and confirm the change was reflected.
9. Next, click on the price button, and change the `period` field. Navigate back
10. Check the `length` of the subscription is now changed to "Never expire"

**Note**: clicking on save or update would reset the data, as we don't send it to the backend yet.

### Images/gif

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt [hunk 1]
```diff
@@ -9,7 +9,7 @@ data class SubscriptionDetails(
     val price: BigDecimal,
     val period: SubscriptionPeriod,
     val periodInterval: Int,
-    val length: Int?,
+    val length: Int = -1,
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt]: Two questions here:
1. Why `-1` instead of `null`?
2. Is the default value intended or a leftover from our discussion on Friday about this topic?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/DropDown.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun WcExposedDropDown(
+    items: List<String>,
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/DropDown.kt]: Just a np, and not necessarly on this PR, WDYT about making this function generic? something like the `setup` function we have for `WCMaterialOutlinedSpinnerView` https://github.com/woocommerce/woocommerce-android/blob/aa4f7b4e8b5741e137537ac2b397e086fa0b6f06/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedSpinnerView.kt#L71-L75

I may be biased since I added it (😅), but I like how it makes dealing with typed values easier.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt [hunk 3]
```diff
@@ -255,7 +258,8 @@ class ProductDetailCardBuilder(
             type = SECONDARY,
             properties = listOf(
                 if (viewModel.isProductUnderCreation) null else product.productReviews(),
-                if (FeatureFlag.PRODUCT_SUBSCRIPTIONS.isEnabled()) product.price() else null,
+                if (isEligibleForSubscriptions()) product.price() else null,
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt]: Why this change? 🙂 
I think we should use only the feature flag and delete it after releasing the feature, because this code is called only for products with the type `subscription`, and adding an additional call that checks the DB is just an overkill that doesn't bring any value, WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetailsMapper.kt [hunk 4]
```diff
@@ -31,8 +31,7 @@ object SubscriptionDetailsMapper {
             val periodInterval = periodIntervalString.toIntOrNull() ?: 0
 
             val lengthString = subscriptionInformation[SubscriptionMetadataKeys.SUBSCRIPTION_LENGTH] ?: ""
-            val lengthInt = lengthString.toIntOrNull()
-            val length = if (lengthInt != null && lengthInt > 0) lengthInt else null
+            val lengthInt = lengthString.toIntOrNull() ?: -1
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetailsMapper.kt]: What if we get a negative value from the API? I know the UI doesn't allow it, but since this WordPress, everything can happen 😄
That's why I think using `null` makes more sense than `-1`, but may be I'm missing something here.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 5]
```diff
@@ -766,48 +792,49 @@ class ProductDetailViewModel @Inject constructor(
         var eventName: AnalyticsEvent? = null
         var hasChanges = false
         when (event) {
-            is ProductExitEvent.ExitSettings -> {
+            is ExitSettings -> {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: np, I think the scoping before the change was more easy to follow, was this changed by the IDE accidentally?

**Review by hichamboushaba — APPROVED** [review]
Nice work @JorgeMucientes, this works well.

I left some comments, I'm pre-approving, but please check them before merging, especially the one about the usage of `isEligibleForSubscription` to control the visibility of rows, and the change to `-1` instead of `null` to handle empty values.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt (hunk 1 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt]: Hey @hichamboushaba it wasn't a leftover. It was something I added after our discussion. The reason I added -1 is because we are treating `null` as the equivalent for `Never expire` and it felt counter intuitive. Not that `-1` is more intuitive but at least is a real value and would allow us to get rid of a few nullchecks here and there. However I had not though of this: 

> What if we get a negative value from the API? I know the UI doesn't allow it, but since this WordPress, everything can happen 😄

Anyway, let's keep it nullable. I've reverted the changes.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/DropDown.kt (hunk 2 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/DropDown.kt]: Nice, I like it. Done.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt (hunk 3 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt]: Hmm, you're right. I used this because I though it would keep the usages of the `FeatureFlag.PRODUCT_SUBSCRIPTIONS` in a single place, but didn't though of the unnecessary DB request we'll be doing each time. Reverted! 👍🏼

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt (hunk 5 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: hmm, I definitely didn't apply this intentionally. I am not sure why the IDE did this 🤔. Reverted

**Comment by JorgeMucientes** [comment]
@hichamboushaba Thanks for the review! I think I've applied all the suggestion 👍🏼. I'll leave some time in case you want to review the latest changes 🙂. If not feel free to merge.

**Comment by JorgeMucientes** [comment]
Hey @hichamboushaba thanks for the review and attention to detail. I've applied all your suggestions. However, I messed up with the 2 branches I had for splitting the work for this task. 
I accidentally rebased `feature/10123-add-subscription-expire-section` on `feature/10123-add-subscription-expire-for-variations` instead of the opposite, which is what I intended  in my head 🤦🏼. Mental note, not to do this at almost 12am 😅. 

Anyway, is not a big deal. This PR now includes the changes to support editing the subscription `length` for product variations. So, to test just following these instructions: 

Variable Subscription
1. Make sure that the Woo Subscriptions plugin is installed in your store.
2. Create a Variable Subscription product, then create a variation (or more).
3. Open the app, then navigate to this product, then open one of the variations.
4. Click on “Subscription expiration” row
5. Change the subscription length.
6. Go back to the product, and confirm the change was reflected.

The most relevant commits are: 
- https://github.com/woocommerce/woocommerce-android/pull/10147/commits/405a97d1fa95e47ad9bdbefcb89a747fc4f9b65e
- https://github.com/woocommerce/woocommerce-android/pull/10147/commits/262b19c1b2afd7fcea0d03aafb15b0bd08a8ae0c


Ready for another round. So sorry for all the mess and inconvenience.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt [hunk 6]
```diff
@@ -829,6 +832,26 @@ class ProductDetailCardBuilder(
...
+                properties = mapOf(
+                    resources.getString(string.subscription_expire) to subscription.expirationDisplayValue(
+                        resources
+                    )
+                ),
+                showTitle = true,
+                onClick = {
+                    viewModel.onEditProductCardClicked(
+                        ViewProductSubscriptionExpiration(subscription),
+                        AnalyticsEvent.PRODUCT_DETAILS_VIEW_SUBSCRIPTIONS_TAPPED
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt]: I just noticed this Tracks event, I think we should remove it, and add a `TODO` comment instead, WDYT?

**Review by hichamboushaba — APPROVED** [review]
Nice work @JorgeMucientes, thank you for making the changes.

I left just one comment about the usage of the Analytics event `PRODUCT_DETAILS_VIEW_SUBSCRIPTIONS_TAPPED` for the expiration, please check it out.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt (hunk 6 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt]: Good catch @hichamboushaba. Removed the tracking and set an explicit reminder in the GH issue https://github.com/woocommerce/woocommerce-android/issues/10163 about adding tracking for both simple and variation products (despite we have a `product_type` property value which will already remind us of it)
